### PR TITLE
Headings outline for markdown

### PR DESCRIPTION
**Pull request by jrieken**
Adds a DocumentSymbolProvider for markdown that lists headings. Should make navigation is large documents easier. 

screen shot 2016-10-05 at 11 04 07

### CODE REVIEW

File: extensions/markdown/src/extension.ts [hunk 1]
```diff
@@ -265,4 +266,33 @@ class MDDocumentContentProvider implements TextDocumentContentProvider {
 			}, 300);
 		}
 	}
+}
+
+class DocumentHeadingsProvider implements vscode.DocumentSymbolProvider {
+
+	// http://daringfireball.net/projects/markdown/syntax#header
+	private static _atxPattern = /^(#){1,6}\s+.+(\s+\1)?/;
```
Comment (mjbvz) [inline extensions/markdown/src/extension.ts]: On the regex:
* Strip out any number of trailing `#`s at the end of the line. This does not have to match the number at the start: http://spec.commonmark.org/0.26/#example-42
* Match end of line too, to handle weird cases like `# a # a #`
* Strip out whitespace around the header text itself for the displayed symbol.

Something like: `^(#{1,6})\s+(.+?)\s*#*\s*$` and use these two captures to build up the symbol

File: extensions/markdown/src/extension.ts [hunk 2]
```diff
@@ -265,4 +266,33 @@ class MDDocumentContentProvider implements TextDocumentContentProvider {
...
+	private static _settext = /^\s*[-=]+\s*$/;
```
Comment (mjbvz) [inline extensions/markdown/src/extension.ts]: Make sure to also test that this does not pick up separators

File: extensions/markdown/src/extension.ts (hunk 1 above)
Comment (mjbvz) [inline extensions/markdown/src/extension.ts]: Also, there are a few other cases to consider, such header like text inside of code blocks. These should not be picked up. Probably not essential for a first pass, but worth considering.

Comment (mjbvz) [inline extensions/markdown/src/extension.ts]: And headers can also indented up to three spaces for each indentation level. I'd make sure to handle the common case, a top level indent of between zero and three spaces (at four it becomes a code block). Handling headers in lists is probably not part of this feature.

**Review by mjbvz — CHANGES_REQUESTED** [review]

**Comment by jrieken** [comment]
@mjbvz I am aware of how this should work in a perfect world using a syntax tree with all bells and whistles. I have no intentions of implementing that 😄 Closing this PR.

**Comment by mjbvz** [comment]
@jrieken I'm not opposed to taking this change in with these limitations (although I would like to see the few suggested fixes for the regular expressions). It would be beneficial to include this in the builtin markdown functionality, even if it it's not perfect.

**Pull request closed without merging**